### ISSUE TITLE
Failed to mount sysext (type overlay) on /run/systemd/sysext/overlay/usr: Too many levels of symbolic links

### ISSUE THREAD

**Issue by stuartpb** [systemd/systemd#26947]
### systemd version the issue has been seen with

253

### Used distribution

Arch

### Linux kernel version used

6.2.7-arch1-1

### CPU architectures issue was seen on

x86_64

### Component

systemd-sysext

### Expected behaviour you didn't see

Looking to noninvasively shadow the default icons that come with code-server, I set up this sysext tree:

```
[drwxr-xr-x root     root    ]  /usr/local/lib/extensions/drafts-icon/
└── [drwxr-xr-x root     root    ]  usr
    └── [drwxr-xr-x root     root    ]  lib
        ├── [drwxr-xr-x root     root    ]  code-server
        │   └── [drwxr-xr-x root     root    ]  src
        │       └── [drwxr-xr-x root     root    ]  browser
        │           └── [drwxr-xr-x root     root    ]  media
        │               ├── [-rw-r--r-- root     root    ]  favicon-dark-support.svg
        │               ├── [-rw-r--r-- root     root    ]  favicon.ico
        │               ├── [-rw-r--r-- root     root    ]  favicon.svg
        │               ├── [-rw-r--r-- root     root    ]  pwa-icon-192.png
        │               ├── [-rw-r--r-- root     root    ]  pwa-icon-512.png
        │               └── [-rw-r--r-- root     root    ]  pwa-icon.png
        └── [drwxr-xr-x root     root    ]  extension-release.d
            └── [-rw-r--r-- root     root    ]  extension-release.drafts-icon
```

The only content of `extension-release.drafts-icon` is `ID=_any`.

I tried `sudo systemctl restart systemd-sysext`, but the overlay was not applied.

### Unexpected behaviour you saw

Opening `journalctl -u systemd-sysext.service`, I find this story:

```
Mar 23 12:52:28 dexter systemd[1]: Starting Merge System Extension Images into /usr/ and /opt/...
Mar 23 12:52:28 dexter (sd-sysext)[316002]: Using extensions 'drafts-icon'.
Mar 23 12:52:28 dexter (sd-sysext)[316002]: Failed to mount sysext (type overlay) on /run/systemd/sysext/overlay/usr (MS_RDONLY "lowerdir=/run/systemd/sysext/meta/usr:/run/systemd/sysext/extensions/drafts-icon/usr:/usr"):
Too many levels of symbolic links
Mar 23 12:52:28 dexter systemd[1]: Finished Merge System Extension Images into /usr/ and /opt/.
```

Even restarting the machine did not resolve the issue.

### Steps to reproduce the problem

Construct an overlay of the described structure, and attempt to apply it on a server with code-server installed and running.

**Comment by poettering** [comment]
is /usr a symlink or something like that?

**Comment by bluca** [comment]
is the sysext a directory store under /usr/ by any chance?

**Comment by stuartpb** [comment]
> is the sysext a directory store under /usr/ by any chance?

Well, yeah, that's what I was indicating with the `tree -pug /usr/local/lib/extensions/drafts-icon/` output above: that this is a directory within `/usr/local/lib/extensions` that describes how to lay out one directory (as well as `/usr/lib/extension-release.d`) under `/usr` (neither of which are `/usr/local`, so I wouldn't expect this to be a problem, though now that I think about how sysext mounts `/usr` and `/opt` [I see how that could happen](https://en.wikipedia.org/wiki/Droste_effect)).

**Comment by stuartpb** [comment]
Can confirm that this does work after `mv /usr/local/lib/extensions/drafts-icon/ /tmp/drafts-icon && mkfs.erofs /usr/local/lib/extensions/drafts-icon.raw /tmp/drafts-icon` as root.

Seems to me that systemd-sysext should either use a strategy that doesn't involve shadowing `/usr/local` when loading from `/usr/lib/extensions/` or `/usr/local/lib/extensions/` (like adding overlay mounts for individual subdirectories), or at least reject this case with a clear explanation (with the documentation noting this limitation).

**Comment by bluca** [comment]
Yeah for the directories case we should definitely detect and reject clearly. You can use `/var/lib/extensions/` or `/etc/extensions/` for storing the extension directory, instead of `/usr/local`, that should get you going.

To be honest I'm not 100% sure why we allow storage for extending /usr under /usr, it does seem a bit strange to me.

**Comment by stuartpb** [comment]
> To be honest I'm not 100% sure why we allow storage for extending /usr under /usr, it does seem a bit strange to me.

It makes sense to me: `/usr/lib/extensions` is for when the distro's packaging model uses some concept of extensions, and `/usr/local/lib/extensions` is for when the "end user's packaging model" uses some concept of extensions. `/etc/extensions` are for extensions that are considered a kind of "configuration", and `/var/lib/extensions` is for "everything else".

**Comment by stuartpb** [comment]
Ahh... I just noticed that sysext makes your `/usr` read-only, so I get what you mean about it being "strange" (since it means the system must "unmerge" in order to modify any of these extensions).

That said, I can still see it making sense, in some contexts (mostly a distro "adopting an extension" while they figure out how to modify the existing packaging infrastructure to incorporate the extension's changes).
